File: aoc/2021/day4/A.cpp
#include <bits/stdc++.h>

// #define ONLINE_JUDGE

#ifndef ONLINE_JUDGE
#include <prettyprint/prettyprint.hpp>
#endif

#define FOR(i, n) for(int i = 0; i < n; ++i)
#define FOR1(i, n) for(int i = 1; i <= n; ++i)
#define TEST int T; std::cin >> T; while(T--)
#define CASE int C; std::cin >> C; FOR1(case_num, C)
#define ALL(v) v.begin(), v.end()
using ll = long long;
using ull = unsigned long long;
using vi = std::vector<int>;
using vll = std::vector<ll>;
using vull = std::vector<ull>;
using pii = std::pair<int, int>;
using pll = std::pair<ll, ll>;
using namespace std;

vector<string> split(const string& str, char delim) {
    istringstream ss(str);
    string token;
    vector<string> tokens;

    while (getline(ss, token, delim)) {
        tokens.push_back(token);
    }

    return tokens;
}

const int BOARD_SIZE = 5;

using Board = array<array<string, BOARD_SIZE>, BOARD_SIZE>;

class BingoBoard {
    unordered_map<string, vector<pair<int, int>>> value_idx;
    array<int, BOARD_SIZE> row_count{}, col_count{};

public:
    BingoBoard(const Board& board) {
        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                value_idx[board[i][j]].emplace_back(i, j);
            }
        }
    }

    void mark_number(string num) {
        if (value_idx.count(num) == 0) return;

        for (auto [i, j] : value_idx.at(num)) {
            ++row_count[i];
            ++col_count[j];
        }

        value_idx.erase(num);
    }

    bool is_won() {
        return any_of(begin(row_count), end(row_count), [](int cnt) {
            return cnt == BOARD_SIZE;
        }) || any_of(begin(col_count), end(col_count), [](int cnt) {
            return cnt == BOARD_SIZE;
        });
    }

    int sum() {
        return accumulate(begin(value_idx), end(value_idx), 0, [](int s, const pair<string, vector<pair<int, int>>>& p) {
            return s + stoi(p.first) * p.second.size();
        });
    }
};

int main() {
   ios_base::sync_with_stdio(false);
   cin.tie(nullptr);
#ifndef ONLINE_JUDGE
    ifstream fin("in.txt");
    ofstream fout("out.txt");
    auto cinbuf = cin.rdbuf(fin.rdbuf());    // save and redirect
    auto coutbuf = cout.rdbuf(fout.rdbuf()); // save and redirect
#endif     
    string numbers_input; cin >> numbers_input;
    vector<string> numbers = split(numbers_input, ',');

    int best_iters = 0, score;

    bool done = false;

    while (!done) {
        Board board;

        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                string elem;
                if (!(cin >> elem)) break;
                board[i][j] = elem;
            }
        }

        if (!cin) {
            done = true;
            break;
        }

        int seen = 0;
        BingoBoard bingo(board);

        auto next = [&]() {
            auto cur = numbers[seen];
            bingo.mark_number(cur);

            return !bingo.is_won();
        };

        while (seen < numbers.size() && next()) ++seen;

        if (best_iters < seen) {
            best_iters = seen;
            int last = stoi(numbers[seen]);
            int sum = bingo.sum();
            score = sum * last;
        }
    }

    cout << score  << endl;
#ifndef ONLINE_JUDGE
    cin.rdbuf(cinbuf);    // restore
    cout.rdbuf(coutbuf); // restore
#endif     
}

File: aoc/2021/day4/A.py
#!/usr/bin/env python3
# encoding: utf-8
import sys

from typing import List

REDIRECT = True
old_stdin = sys.stdin

if REDIRECT:
    sys.stdin = open("in.txt", "r")


numbers_input = input()
numbers = [int(x) for x in numbers_input.split(',')]

best_seen = len(numbers)
score = 0

BOARD_SIZE = 5


class BingoBoard:
    def __init__(self, board: List[List[int]]):
        self._val_idx = {}
        self._row_counts = [0 for _ in range(BOARD_SIZE)]
        self._col_counts = [0 for _ in range(BOARD_SIZE)]

        assert(len(board) == BOARD_SIZE)

        for i in range(BOARD_SIZE):
            for j in range(BOARD_SIZE):
                val = board[i][j]
                idx = self._val_idx.get(val, [])
                idx.append((i, j))
                self._val_idx[val] = idx

    def get_sum(self) -> int:
        return sum([k * len(v) for k, v in self._val_idx.items()])

    def is_won(self):
        return BOARD_SIZE in self._row_counts or BOARD_SIZE in self._col_counts

    def mark_number(self, number: int):
        if number not in self._val_idx:
            return

        for i, j in self._val_idx[number]:
            self._row_counts[i] += 1
            self._col_counts[j] += 1

        del self._val_idx[number]


while True:
    # Read the 5 lines of the board.
    board = []

    for line in sys.stdin:
        line = line.strip()

        if not line:
            continue

        board.append([int(x) for x in line.split()])

        if len(board) == BOARD_SIZE:
            break

    if len(board) != BOARD_SIZE:
        break

    bingo = BingoBoard(board)
    seen = 0

    while seen < len(numbers) and not bingo.is_won():
        bingo.mark_number(numbers[seen])
        seen += 1
    
    if seen < best_seen:
        best_seen = seen
        last = numbers[seen - 1]
        bingo_sum = bingo.get_sum()
        score = bingo_sum * last

print(score)

if REDIRECT:
    sys.stdin.close()
    sys.stdin = old_stdin